segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,336 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        return this.taskScheduler.add(Arrays.asList(keyNames), () -> executeExclusive(operation, keyNames));
+    }
 
-        // Find chunks to delete.
-        val chunksToDelete = new ArrayList<GarbageChunkInfo>();
-        int count = 0;
+    /**
+     * Executes the given Callable asynchronously and exclusively.
+     * It returns a CompletableFuture that will be completed with the result.
+     * The operations are not allowed to be concurrent.
+     *
+     * @param operation    The Callable to execute.
+     * @param <R>       Return type of the operation.
+     * @param keyNames The names of the keys involved in this operation (for sequencing purposes).
+     * @return A CompletableFuture that, when completed, will contain the result of the operation.
+     * If the operation failed, it will contain the cause of the failure.
+     * */
+    private <R> CompletableFuture<R> executeExclusive(Callable<CompletableFuture<R>> operation, String... keyNames) {
+        return CompletableFuture.completedFuture(null).thenComposeAsync(v -> {
+            Exceptions.checkNotClosed(this.closed.get(), this);
+            try {
+                return operation.call();
+            } catch (CompletionException e) {
+                throw new CompletionException(Exceptions.unwrap(e));
+            } catch (Exception e) {
+                throw new CompletionException(e);
+            }
+        }, this.storageExecutor);
+    }
 
-        // Wait until you have at least one item or timeout expires.
-        GarbageChunkInfo info = Exceptions.handleInterruptedCall(() -> garbageChunks.poll(config.getGarbageCollectionDelay().toMillis(), TimeUnit.MILLISECONDS));
-        log.trace("{}: deleteGarbage - retrieved {}", traceObjectId, info);
-        while (null != info ) {
-            queueSize.decrementAndGet();
-            chunksToDelete.add(info);
 
-            count++;
-            if (count >= maxItems) {
-                break;
-            }
-            // Do not block
-            info = garbageChunks.poll();
-            log.trace("{}: deleteGarbage - retrieved {}", traceObjectId, info);
+    private CompletableFuture<Void> processTask(TaskInfo infoToDelete) {
+        if (infoToDelete.taskType == TaskInfo.DELETE_CHUNK) {
+            return deleteChunk(infoToDelete);
         }
-
-        // Sleep if no chunks to delete.
-        if (count == 0) {
-            log.debug("{}: deleteGarbage - no work - sleeping for {}.", traceObjectId, config.getGarbageCollectionDelay());
-            return CompletableFuture.completedFuture(false);
+        if (infoToDelete.taskType == TaskInfo.DELETE_SEGMENT) {
+            return  deleteSegment(infoToDelete);
+        }
+        if (infoToDelete.taskType == TaskInfo.DELETE_JOURNAL) {
+            return  CompletableFuture.completedFuture(null);
         }
+        return null;
```

Comment:
I would prefer to log a WARN if we get here. Throwing might be too extreme as we may have problems when we do upgrades and add new types of tasks. But we should at least know about them.

Reply:
Fixed.